### PR TITLE
[Hack] Add Semgrep extension tests

### PR DESCRIPTION
**Pull request by frankeld**
~⚠️  BUILD WILL FAIL (requires updated semgrep-hack build - see https://github.com/returntocorp/ocaml-tree-sitter-semgrep/pull/244; once semgrep-hack is updated, I will add another commit to fix the submodule version)~ Done
- Adds many test cases for Semgrep matching
- Updates Parse_hack to support Semgrep extensions
- Misc. parse file updates (type params)

Following https://semgrep.dev/docs/language-support/#maturity-definitions, I think we have reached alpha. If so, I can open another PR to add Hack where relevant. 

PR checklist:
- [ ] documentation is up to date
- [ ] changelog is up to date

### CODE REVIEW

**Comment by frankeld** [comment]
@mjambon I anticipate the `semgrep unit and e2e tests` failure is because Hack hasn't been added as a supported language in certain files. I will follow along with the python file changes I found [here](https://github.com/returntocorp/semgrep/pull/3167) unless you have any other suggestions.

File: semgrep/semgrep/target_manager_extensions.py [hunk 1]
```diff
@@ -42,6 +43,7 @@
     + C_EXTENSIONS
     + GO_EXTENSIONS
     + RUBY_EXTENSIONS
```
Comment (frankeld) [inline semgrep/semgrep/target_manager_extensions.py:45]: Why isn't PHP here? Are only certain languages added?

Comment (mjambon) [reply semgrep/semgrep/target_manager_extensions.py:45]: I don't know either. This file needs some clean-up:
* It looks like `ALL_EXTENSIONS` should be derived from one of the dicts defined below anyway.
* The languages should be ordered alphabetically to make it easier to find a specific language from those lists.

I'd say don't worry about it.

**Review by mjambon — APPROVED** [review]
I didn't look into the details. Looks good to me!

**Pull request merged**